sync/upstream.py

```diff
@@ -462,6 +462,83 @@ def pr_head(self):
         ref = self.git_wpt.refs[pr_ref]
         return ref.commit.hexsha
 
+    @property
+    def pr_commits(self):
+        pr_head = self.pr_head
+        if not pr_head:
+            logger.error(
+                "Can't get PR commits as the ref head could not be found for %s" %
+                self.process_name
+            )
+            return
+        else:
+            pr_head = self.git_wpt.commit(pr_head)
+
+        pr = None
+
+        # Retrieve the commit that pushed these changes to master
+        tag = 'merge_pr_%d' % self.pr
+        if tag in self.git_wpt.tags:
+            merge_commit_sha = self.git_wpt.tags[tag].commit
+        elif self.status in ('wpt-merged', 'complete'):
+            logger.warning("Merge Tag not found for %s" % tag)
+            pr = env.gh_wpt.get_pull(self.pr)
+            merge_commit_sha = pr["merge_commit_sha"]
+        elif self.pr_status == 'open':
+            # PR must still be open, just use master as our 'merge commit'
+            merge_commit_sha = self.git_wpt.head.commit.hexsha
+        else:
+            logger.error("Couldn't determine state of the PR")
+            return
+
+        merge_commit = self.git_wpt.commit(merge_commit_sha)
+
+        pr_head_reachable = self.git_wpt.is_ancestor(pr_head.hexsha, merge_commit_sha)
+
+        # IF PR head reachable, scenario is either a normal merge or rebase + fast forward
+        if pr_head_reachable:
+            # If the commit has two parents, one of them being our pr head, it is a merge commit
+            if len(merge_commit.parents) == 2 and pr_head in merge_commit.parents:
+                parents = list(merge_commit.parents)
+                other_parent = parents[0] if parents[1] == pr_head else parents[1]
+                merge_base = self.git_wpt.merge_base(pr_head, other_parent)[0]
+
+            # Singular parent, fast forward merge
+            elif len(merge_commit.parents) == 1 and merge_commit == pr_head:
+
+                # Fall back to the API for this
+                if not pr:
+                    pr = env.gh_wpt.get_pull(self.pr)
+                merge_base = self.git_wpt.commit(pr["base"]["sha"])
+
+        # Squash and Merge
+        # If the pr head is not reachable from master then it must be a typical squash merge
```

Comment:
Oh I didn't realise this could occur, I was under the impression the entire PR would rebase and move to M2 as the merge base. Thanks for pointing this out.